vector/src/Data/Vector/Fusion/Bundle/Monadic.hs

```diff
@@ -1091,9 +1092,10 @@ fromVector v = v `seq` n `seq` Bundle (Stream step 0)
     n = basicLength v
 
     {-# INLINE step #-}
-    step i | i >= n = return Done
-           | otherwise = case basicUnsafeIndexM v i of
-                           Box x -> return $ Yield x (i+1)
+    step (I# i)
+      | I# i >= n = return Done
```

Comment:
Here and everywhere in this PR, when we pattern-match on `I#` but then construct it again in function's body is GHC 100% guaranteed to optimise allocation of fresh `Int` away and will just reuse whatever we pattern-matched on?

Reply:
GHC _should_ be quite good at this. So it should eliminate `Int` allocations.